src/osd/PGLog.h

```diff
@@ -1720,6 +1716,73 @@ struct PGLog : DoutPrefixProvider {
       store, ch, info, log, missing, pgmeta_oid, dpp})->start();
   }
 
+  seastar::future<> rebuild_missing_set_with_deletes(
+    crimson::os::FuturizedStore *store,
+    crimson::os::CollectionRef ch,
+    const pg_info_t &info)
+  {
+    return seastar::do_with(map<hobject_t, pg_missing_item>(),
+      [this, info, store, ch] (auto& extra_missing) {
+      // save entries not generated from the current log (e.g. added due
+      // to repair, EIO handling, or divergent_priors).
+      for (const auto& p : missing.get_items()) {
+	if (!log.logged_object(p.first)) {
+	  extra_missing[p.first] = p.second;
+	}
+      }
+      missing.clear();
+
+      // go through the log and add items that are not present or older
+      // versions on disk, just as if we were reading the log + metadata
+      // off disk originally
+      return seastar::do_with(set<hobject_t>(), log.log.rbegin(),
+	[this, info, store, ch, &extra_missing] (auto& did, auto& i) {
+	return seastar::repeat([this, &did, &i, info, store, ch] {
+	  if (i == log.log.rend() || i->version <= info.last_complete) {
+	    return seastar::make_ready_future
+			<seastar::bool_class<seastar::stop_iteration_tag>>
+			(seastar::stop_iteration::yes);
+	  }
+	  if (i->soid > info.last_backfill ||
+	      i->is_error() ||
+	      did.find(i->soid) != did.end()) {
```

Comment:
```suggestion
	  if (i->soid > info.last_backfill ||
	      i->is_error()) {
	    return seastar::make_ready_future<seastar::stop_iteration>(
	      seastar::stop_iteration::no);
	  }
	  auto [it, added] = did.emplace(i->soid);
	  if (!added) {
	    return seastar::make_ready_future<seastar::stop_iteration>(
	      seastar::stop_iteration::no);
	  }
	  // check oi.version
```

please note, find() in a container and then insert into it is an anti-pattern when using STL container, it hurts the performance and readability.